File: lintcode/python/0726_check_full_binary_tree.py
#!/usr/bin/python -t

# BFS

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

from collections import deque

class Solution:
    """
    @param root: the given tree
    @return: Whether it is a full tree
    """
    def isFullTree(self, root):
        # write your code here
        if not root:
            return True
            
        q = deque([root])
        
        while len(q) > 0:
            l = len(q)
            
            for i in range(l):
                node = q.popleft()
                if (node.left and not node.right) or (not node.left and node.right):
                    return False
                
                if node.left and node.right:
                    q.append(node.left)
                    q.append(node.right)
                    
        return True
        

File: lintcode/golang/0726_check_full_binary_tree.go


// bfs

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the given tree
 * @return: Whether it is a full tree
 */
func isFullTree (root *TreeNode) bool {
    // write your code here
    if root == nil {
        return true
    }
    
    q := make([]*TreeNode, 0)
    q = append(q, root)
    
    for len(q) != 0 {
        level := make([]*TreeNode, 0)
        for _, node := range q {
            if (node.Left == nil && node.Right != nil) || (node.Left != nil && node.Right == nil) {
                return false
            }
            if node.Left != nil {
                level = append(level, node.Left)
            }
            if node.Right != nil {
                level = append(level, node.Right)
            }
        }
        q = level
    }
    
    return true
}

// recursion


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the given tree
 * @return: Whether it is a full tree
 */
func isFullTree (root *TreeNode) bool {
    // write your code here
    if root == nil {
        return true
    }
    
    if root.Left == nil && root.Right == nil {
        return true
    }
    
    if root.Left == nil || root.Right == nil {
        return false
    }
    
    return isFullTree(root.Left) && isFullTree(root.Right)
}

